components/engine_rocks_helper/src/sst_recovery.rs

```diff
@@ -0,0 +1,265 @@
+                        ),
+                    );
+                }
+
+                // lock the store_meta and check if the range exist.
+                let _ = self.check_overlap_damaged_regions(f);
+
+                return;
+            }
+        }
+    }
+
+    // `sst_path` has been processed and is still in a damaged state.
+    fn exist_scheduling_regions(&self, sst_path: &str) -> bool {
+        self.damaged_files.iter().any(|f| f.name == sst_path)
+    }
+
+    // Periodically check for recorded damaged files。
+    //
+    // Acquire meta lock.
+    fn check_damaged_files(&mut self) {
+        let mut new_damaged_files = self.damaged_files.clone();
+        new_damaged_files.retain(|f| {
+            if f.start_time.elapsed() > self.max_hang_duration {
+                self.set_panic_mark_and_panic(
+                    &f.name,
+                    &format!("recovery job exceeded {:?}", self.max_hang_duration),
+                );
+            }
+            self.check_overlap_damaged_regions(f.clone())
+        });
+        TIKV_ROCKSDB_DAMAGED_FILES.set(new_damaged_files.len() as i64);
+        self.damaged_files = new_damaged_files;
+    }
+
+    // Check whether the StoreMeta contains the region range, if it contains,
+    // recorded fault region ids to report to PD and add file info into `damaged_files`.
+    //
+    // Acquire meta lock.
+    fn check_overlap_damaged_regions(&mut self, file: FileInfo) -> bool {
+        let mut meta = self.store_meta.lock().unwrap();
+
+        // Find overlapping region ids.
+        // Condition:
+        // end_key > file.smallestkey
+        // start_key <= file.largestkey
+        let mut ids = vec![];
+        for (_, id) in meta
+            .region_ranges
+            .range((Excluded(file.smallest_key.clone()), Unbounded::<Vec<u8>>))
+        {
+            let region = &meta.regions[id];
+            if keys::enc_start_key(region) <= file.largest_key {
+                ids.push(*id);
+            }
+        }
+
+        let should_overlap = || {
+            fail_point!("sst_recovery_should_overlap", |t| -> Option<bool> {
+                Some(t.unwrap().parse::<bool>().unwrap())
+            });
+            None
+        };
+        let overlap = should_overlap().unwrap_or(!ids.is_empty());
+
+        if overlap {
+            // damaged file context should be updated.
+            for id in ids {
+                meta.damaged_regions_id.insert(id);
+            }
+            if !self.exist_scheduling_regions(&file.name) {
+                self.damaged_files.push(file);
+                TIKV_ROCKSDB_DAMAGED_FILES.inc();
+            }
+        } else {
+            fail_point!("sst_recovery_before_delete_files");
+            // The sst file can be deleted safely.
+            // set `include_end` to `true` otherwise the file with the same largest key will be skipped.
+            self.db
+                .delete_files_in_range(&file.smallest_key, &file.largest_key, true)
```

Comment:
Not good to do this under lock?

Reply:
it's necessary to keep correctness considering concurrent add-peer. Should add comment here to explain

Reply:
Added comment

Reply:
So a peer cannot be added when the data is missing? What is the exact row of data that determines this logic?

Reply:
a peer with same range can be added back when the data is missing but cannot when cleanup logic is executing.

Reply:
So this lock is to prevent newly added peer mistakenly deleted here?

Reply:
Yes, it's about to clean up the range. If there is an add-peer command add peer on this TiKV and apply a snapshot concurrently. This clean-up may delete the data of the newly added peer mistakenly.

Reply:
"So this lock is to prevent newly added peer mistakenly deleted here?" 
No, the peer can't be added due to lock is hold.